jazzy-elks/dummy_express
# Jetstream subscriber

Current status of the code:

The edge acts as a push subscriber, meaning it will wait for the publisher to publish a message, and then acknowledge the receipt of that message.

The jetstream code has been extracted into a `jetstream.js` file in the `lib` directory.

In `bin/www` the subscription code is initialized on line 46 after the server is listening via`  subscription.initSubscribe();`. The rest of the relevant code is in `lib/jetstream.js`.

Within `jetstream.js`, we currently have three functions `updateSubscribe()`, `createSubscribe()`, and `deleteSubscribe()`.  Each of these functions subscribes to the relevant subject.  Common/repeated code was extracted into the `subscriptionLog()` and `config()` functions, which act as helper functions.

To test/see this code in action:

start the application with `npm start`.  You will see:

```
> dummy-express@0.0.0 start
> node ./bin/www

 >> jetstream functionality in nats.js is preview functionality 
Update subscription running
Create subscription running
Delete subscription running
```

Using postman to create a new flag via our monolith's API, we will see the edge server log this message:

```
Data from Create func; subject: FLAG.created: New flag created. Data: {
  toggledOn: false,
  _id: 60df181f50bd260415e84df7,
  title: 'test flag creation subscription',
  createdAt: 2021-07-02T13:43:59.162Z,
  updatedAt: 2021-07-02T13:43:59.162Z,
  __v: 0
}
```

Using postman to edit this same flag, our edge server logs:

```
Data from Update func; subject: FLAG.updated: Flag updated. Data: {
  toggledOn: false,
  _id: 60df181f50bd260415e84df7,
  title: 'test flag creation subscription 2.0',
  createdAt: 2021-07-02T13:43:59.162Z,
  updatedAt: 2021-07-02T13:44:51.825Z,
  __v: 0
}
```

Using postman to delete this flag:

```
Data from Delete func; subject: FLAG.deleted: Flag deleted.
```

---

Another interesting test to see our subscription working:

1. Comment out line 54 of `lib/jetstream.js` so that the `updateSubscribe()` function is never invoked, meaning no 'update' messages can be received.

2. Start the edge server and then edit an existing flag via PUT using postman.  You'll notice no message receipt is logged to the edge's server.

3. Stop the edge server, uncomment line 54, then start the edge server again so that `updateSubscribe()` is invoked on startup. Do not submit another PUT request via postman.

4. The message about the update is now received! Edge server log below:
```
Data from Update func; subject: FLAG.updated: Flag updated. Data: {
  toggledOn: false,
  _id: 60da6da3befae7d66f0f807f,
  title: 'test flag creation subscription 2.0',
  createdAt: 2021-06-29T00:47:31.604Z,
  updatedAt: 2021-07-02T13:47:16.439Z,
  __v: 0
}
```
